agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -57,14 +63,52 @@ def deploy_rev_changed?
       @deploy_rev_changed == true
     end
 
-    # @param [String] topic
-    # @param [Docker::Event] e
-    def on_container_event(topic, e)
-      attrs = e.actor.attributes
+    # @param topic [String]
+    # @param event [Docker::Event]
+    def on_container_event(topic, event)
+      attrs = event.actor.attributes
       if attrs['io.kontena.service.id'] == @service_pod.service_id &&
+          attrs['io.kontena.container.type'] == 'container'.freeze &&
           attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number
         @container_state_changed = true
+        handle_restart_on_die if event.status == 'die'.freeze
```

Comment:
Separate issue for fixing the `on_container_event` to ignore old events, and not think that the container died and needs to be restarted immediately after each redeploy: #2766